Product Collection - Add Inherit query from template control

kmanijak opened the pull request:
> ⚠️ This PR will be merged after PR #9634, as It also contains changes made in #9634

This PR adds a setting to Product Collection to inherit the global query.
<img width="270" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/16707866/815a4676-5f2e-4819-a251-92b5a4d1049f">


When enabled:
- all the filters and ordering options are hidden
- the actual global query should be in use
- make sure Product Collection works fine with the Catalog Sorting and Product Results Count blocks

Fixes https://github.com/woocommerce/woocommerce-blocks/issues/9358

### Testing

#### Testing with Product Archive templates

1. Open the Editor and navigate to the "Product Search Results"
2. Replace the existing 'Classic Template' block with the 'Product Collection' block.
3. After the 'Product Collection' block is successfully inserted, ensure that the 'Inherit query from template' option is enabled by default.
4. Save the changes.
5. Navigate to the frontend of your website. You can emulate a search for "shirt" by visiting the URL "/shop?s=shirt". Confirm that the search query `shirt` effectively displays related products.
6. Head back to the Editor.
7. In the 'Product Collection' block, deactivate the 'Inherit query from template' option.
8. Subsequently, you should observe various settings and filters available in the inspector controls. Modify these filters and settings as required and validate immediately that the changes are visible within the Editor.
9. Save your modifications and return to the frontend of your website.
10. Ensure that the adjustments made to the 'Product Collection' block within the Editor are accurately displayed on the frontend of your website.

Here is a quick demo:

https://github.com/woocommerce/woocommerce-blocks/assets/16707866/5703fa10-897c-4310-8ad1-f9c176014737


#### Testing in Post

1. Create a new post
2. Add Product Collection block to the post
3. With the 'Product Collection' block selected, open the Inspector Controls (located in the right sidebar). Confirm that the option `inherit query from template` does not appear in the list of options. This is because it's relevant only for Product Archive pages and should not be visible in this context.

https://github.com/woocommerce/woocommerce-blocks/assets/16707866/b4dce46a-7853-42f4-ae40-545850d121e9


* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [ ] WooCommerce Core
* [ ] Feature plugin
* [x] Experimental

### Changelog

> Product Collection - Add `Inherit query from template` control

imanish003 committed:
Add columns control to product collection block editor settings

- `InspectorControls` from './inspector-controls' is now imported in `edit.tsx` and used in the returned JSX of `Edit` function.
- A new file `columns-control.tsx` is added under 'product-collection' block's 'inspector-controls' directory which exports a `ColumnsControl` component. This component uses `RangeControl` from '@wordpress/components' to control the number of columns in the product collection display layout when the layout type is 'flex'.
- The types file (`types.ts`) for 'product-collection' block is updated. The `Attributes` interface is renamed to `ProductCollectionAttributes` and the `ProductCollectionContext` interface is removed. The `ProductCollectionAttributes` now includes 'queryContext', 'templateSlug', and 'displayLayout' properties.

imanish003 committed:
Merge branch 'trunk' into 9356-product-collection-editor-settings-columns

imanish003 committed:
Refactor: Simplify Fallback Return in ColumnsControl Component

This commit simplifies the fallback return value of the ColumnsControl component. Instead of returning an empty fragment (<> </>), it now returns null when the condition isn't met. This change improves readability and aligns with best practices for conditional rendering in React.

imanish003 committed:
Feature: Add 'Order By' Control to Product Collection Inspector

This commit adds a new 'Order By' control to the product collection inspector. The control allows users to specify the order of products in a collection by various attributes such as title and date. To support this, a new component 'OrderByControl' has been created and included in the product collection inspector. Additionally, the types for 'order' and 'orderBy' attributes have been updated and exported for reuse.

imanish003 committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into 9359-product-collection-editor-settings-order-by

imanish003 committed:
Add more options to OrderBy type

kmanijak committed:
Add Inherit global query control to the Product Collection

imanish003 committed:
Merge branch 'trunk' into 9359-product-collection-editor-settings-order-by

imanish003 committed:
Add orderby handling on frontend & editor

The main changes include:
1. Added a new property 'isProductCollectionBlock' in the block.json to denote if a block is a product collection block.
2. In the ProductCollection PHP class, a new initialization function has been defined to hook into the WordPress lifecycle, register the block, and update the query based on this block.
3. Added methods to manage query parameters for both frontend rendering and the Editor.
4. Expanded allowed 'collection_params' for the REST API to include custom 'orderby' values.
5. Defined a function to build the query based on block attributes, filters, and global WP_Query.
6. Created utility functions to handle complex query operations such as merging queries, handling custom sort values, and merging arrays recursively.

These improvements allow for more flexible and robust handling of product collections in both the front-end display and the WordPress editor. It also extends support for custom 'orderby' values in the REST API, which allows for more advanced sorting options in product collections.

imanish003 committed:
Merge branch 'trunk' into 9359-product-collection-editor-settings-order-by

imanish003 committed:
Merge branch 'trunk' into 9359-product-collection-editor-settings-order-by

kmanijak committed:
Merge branch 'trunk' into add/product-collection-inherit

kmanijak committed:
Fix the condition to show query controls (they shoul appear if query is NOT inherited)

imanish003 committed:
Merge branch 'trunk' into add/product-collection-inherit

imanish003 reviewed (commented):
Hi @kmanijak, great job as always! 🙌🏻

I have tested the pull request, and it's working as described in the testing steps. 🚀
I wanted to mention that currently, in the Products beta block, we hide the "Inherit query from template" setting when it doesn't make sense, such as when adding the block to a post. Maybe we should do the same for the Product Collection block as well. What do you think? 🙂

> At the moment initial query settings are hardcoded in the Product Collection. Change the default settings, for example by changing the order to desc in [here](https://github.com/woocommerce/woocommerce-blocks/blob/a5d9d5cb16aaced643ff23e42ea716feb896a539/assets/js/blocks/product-collection/block.json#L22)

I merged the trunk with this branch, and now we have an option in the sidebar settings to change the order. Therefore, we won't need to perform this step 🙂

kmanijak committed:
Make Product Collection inheirt global query in product archive templates by default

kmanijak commented:
> I wanted to mention that currently, in the Products beta block, we hide the "Inherit query from template" setting when it doesn't make sense, such as when adding the block to a post. Maybe we should do the same for the Product Collection block as well. What do you think? 🙂

@imanish003, I agree. I added this logic in [this commit](https://github.com/woocommerce/woocommerce-blocks/pull/9485/commits/d257e8bdb014742c40ef069110f6a2a35148fb7a), but there's a problem that I won't have time to fix.

When adding the Product Collection, attributes are correctly set in the Editor (client-side rendered), but when you save with no changes and go to frontend then query is not consumed. This is due to fact that until any change is made in the inspector control and until `setAttributes` function is called the attributes are not saved to the block and hence not read by server-side.

Before changing any setting | After changing setting
--- | ---
<img width="951" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/20098064/3bf0d65d-4a03-42e1-98fd-a45ba40c2830"> | <img width="1014" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/20098064/c70823b8-52b2-4f36-9629-190ce1c85a64">

I tried extending this initial `setAttributes` with the `query` [here](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/assets/js/blocks/product-collection/edit.tsx#L89), but even though everything looked fine I couldn't find why the query attribute is not set. I'm pretty sure it's something dumb and simple, but under time pressure I couldn't find it.

1. I changed the testing steps to include "Order by" setting and the logic that "Inherit..." option is enabled in templates and not disabled in post/page editor
2. I converted this PR to a draft until the initial attributes problem is solved.
3. Like with the other PR, feel free to pick it up or wait until I'm back - depends on the priorities 🙏

imanish003 committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into 9361-product-collection-filters-on-sale

imanish003 committed:
Add 'on sale' filter and enhance settings management in product collection block

This commit introduces several changes to the product collection block.
- First, it adds a new 'on sale' filter that can be used to display only the products that are currently on sale.
- It also refactors the settings management in the product collection block to use the experimental ToolsPanel component from WordPress, which provides a more flexible and intuitive way to manage block settings.
- It moves the 'Columns' control into the ToolsPanel, along with the 'Order by' control.
- A new utility function `setQueryAttribute` is introduced to simplify setting nested query parameters.
- The structure of the `ProductCollectionAttributes` and `ProductCollectionQuery` types have been adjusted to accommodate the changes.
- Finally, it makes corresponding changes in the PHP part to handle the new 'on sale' query parameter.

This should enhance the flexibility and user-friendliness of the product collection block.

imanish003 committed:
Merge branch 'trunk' into 9361-product-collection-filters-on-sale

imanish003 committed:
Merge branch 'trunk' into add/product-collection-inherit

imanish003 committed:
Merge branch 'trunk' into add/product-collection-inherit

imanish003 committed:
Revert "Make Product Collection inheirt global query in product archive templates by default"

This reverts commit d257e8bdb014742c40ef069110f6a2a35148fb7a.

imanish003 committed:
Updated 'inherit' behavior in the Product Collection block

This commit updates the 'inherit' behavior in the Product Collection block and its associated controls.

Changes include:
- Removed the 'inherit' attribute from the block's JSON definition
- Defined an array of 'archive product templates' which includes the WooCommerce product archive, taxonomy, attribute, and search results templates.
- Set the initial 'inherit' value based on the current template ID when the Product Collection block is first added to the page.
- Restored the query object value when toggling 'inherit' off.
- Conditionally rendered the InheritQueryControl based on the current editor being the site editor.

imanish003 committed:
Merge branch 'trunk' into add/product-collection-inherit

imanish003 committed:
Add stock status filter to WooCommerce product collection block

This commit introduces a stock status filter to the WooCommerce product collection block.

The changes include:
1. Added the ability to filter products based on their stock status within the 'product-collection' block. A new stock status control is created within the inspector-controls of the block.
2. A new 'get_stock_status_query' function is introduced in 'ProductCollection.php' which returns a query for products depending on their stock status.

Please note that the stock status filter will only appear in the experimental build for now.

imanish003 committed:
Refactor Stock Status control of Product Collection block

This commit refactors the Stock Status control. The changes aim to improve the code organization and make the behavior of the component more explicit.

The key modifications are:
1. Moved stock status related constants and functions from `inspector-controls/utils.tsx` to `inspector-controls/constants.ts`. This is done to ensure that all constants and similar utility functions are organized in one place.
2. Updated `product-collection/index.tsx` to import `getDefaultStockStatuses` from `inspector-controls/constants` instead of `inspector-controls/utils`.
3. Updated `stock-status-control.tsx` to determine whether the stock status has value or not by comparing with the default stock statuses using `fastDeepEqual`. If the stock status control is deselected, it resets the stock status to the default statuses.

These changes do not introduce any new functionalities, but improve the readability and maintainability of the code.

imanish003 committed:
Merge branch 'trunk' into 9362-product-collection-filters-stock-status

imanish003 committed:
Add keyword search control to Product Collection block

This commit introduces a keyword search functionality to the Product Collection block. The update is aimed to provide users with more flexibility and precision in product collection queries.

Key changes:
1. Introduced a new file `keyword-control.tsx` that creates a Keyword Control component. This component includes a TextControl field that allows inputting a search keyword. The keyword search is debounced to prevent unnecessary queries during input and updates the block's attributes accordingly.
2. Modified `inspector-controls/index.tsx` to include the KeywordControl in the ToolsPanel for the block's filters.
3. Adjusted `ProductCollection.php` to include the keyword search in the product query array.

With these changes, users can now search for products by keyword in the Product Collection block.

imanish003 committed:
Resolve conflicts

imanish003 committed:
Add product attributes filter control to ProductCollection block

- This commit introduces the ability to filter products by attributes in ProductCollection block.
- A new `woocommerceAttributes` key was added to the `block.json` file and the `ProductCollectionQuery` type. Also, a new file `attributes-control.tsx` was created, providing the UI component for the attribute filter control in the editor.
- In addition, updates were made to the `ProductCollection.php` file in the backend to support filtering products by attributes, and the tax query was updated to include attribute queries.
- Lastly, the `ProductCollectionInspectorControls` was updated to include the `AttributesControl` component, thus enabling users to filter products by attributes in the block editor."`

imanish003 committed:
Merge branch 'trunk' into 9363-product-collection-filters-attributes

thealexandrelara committed:
Merge branch 'trunk' into 9363-product-collection-filters-attributes

imanish003 committed:
Merge branch 'trunk' into 9363-product-collection-filters-attributes

imanish003 committed:
Resolve conflicts

imanish003 committed:
remove unused import of getDefaultStockStatuses

imanish003 committed:
Resolve conflicts

imanish003 committed:
Delete a duplicate file

imanish003 committed:
Resolve conflicts

imanish003 committed:
Remove console log

imanish003 committed:
Merge branch 'trunk' into 9363-product-collection-filters-attributes

imanish003 committed:
Add taxonomies control to Product collection block

The primary changes include:
1. `taxQuery` field in the `ProductCollectionAttributes` was changed from a string to an object in `assets/js/blocks/product-collection/types.ts` and `assets/js/blocks/product-collection/constants.ts`, accommodating the ability to query products by taxonomy terms.

2. `assets/js/blocks/product-collection/inspector-controls/utils.tsx` was moved to `assets/js/blocks/product-collection/utils.tsx` to make it available for broader use.

3. New component `TaxonomyControls` was created in `assets/js/blocks/product-collection/inspector-controls/taxonomy-controls.tsx`, which is included in `assets/js/blocks/product-collection/inspector-controls/index.tsx`. This new control allows users to filter products in the block by their taxonomy terms.

4. Updated the block's inspector controls in `assets/js/blocks/product-collection/inspector-controls/index.tsx` to use the new `TaxonomyControls` component.

Please note that the TaxonomyControls component uses experimental features of WordPress's FormTokenField. As a result, a comment has been added to disable eslint warnings regarding the use of experimental APIs.

imanish003 committed:
Merge branch 'trunk' into 9364-product-collection-filters-taxonomies

imanish003 committed:
Address PR feedback & other improvements

1. Added `woocommerceAttributes` to `DEFAULT_FILTERS` in the `product-collection/constants.ts` file to fix `reset all` button issue.

2. Refactored `attributes-control.tsx` to make it more maintainable:
   - The constant `EDIT_ATTRIBUTES_URL` now uses `ADMIN_URL` from `@woocommerce/settings` for a more dynamic URL generation.
   - The interface `Props` has been renamed to `AttributesControlProps` for more explicit naming.
   - Removed the usage of `useState` and `useEffect` for selected attributes. Instead, `selectedAttributes` is now directly derived from `woocommerceAttributes`.
   - The CSS class `woocommerce-product-query-panel__external-link` was renamed to `wc-product-collection-panel__external-link`

3. Deleted the `product-collection/inspector-controls/constants.ts` file which was no longer necessary due to changes in product collection implementation.

These changes contribute to codebase quality, improving readability and maintainability.

imanish003 committed:
Merge branch 'trunk' into 9363-product-collection-filters-attributes

imanish003 committed:
Merge branch '9363-product-collection-filters-attributes' of https://github.com/woocommerce/woocommerce-blocks into 9364-product-collection-filters-taxonomies

imanish003 committed:
Address PR review comments

This commit involves a significant refactoring of the default product query inside the 'product-collection/constants.ts' file. A new constant `DEFAULT_QUERY` has been defined and used to replace the previously hard-coded default query settings. This refactoring aids in code readability and future modifications.

Changes also include adjustments in 'product-collection/inspector-controls' to enhance UI/UX. A new SCSS file 'editor.scss' has been created for custom styles related to the editor.

Additions include:
- Adding a class name `product-collection-inspector-toolspanel__filters` to ToolsPanel for additional styling.
- The experimental property `__experimentalShowHowTo` is set to false for `FormTokenField` and `StockStatusControl`, to hide some additional information.

In 'product-collection/inspector-controls/taxonomy-controls.tsx', the classname `product-collection-inspector__taxonomy-control` is added for improved CSS targeting.

imanish003 committed:
Add wc-block-editor prefix to className

imanish003 committed:
Merge branch 'trunk' into 9363-product-collection-filters-attributes

imanish003 committed:
Merge branch '9363-product-collection-filters-attributes' of https://github.com/woocommerce/woocommerce-blocks into 9364-product-collection-filters-taxonomies

imanish003 committed:
Resolve conflicts

imanish003 committed:
Make improvements to 'inherit' functionality in Product collection block.

Key changes are:

1. 'inherit' in 'ProductCollectionAttributes' within 'constants.ts' has been changed from 'false' to 'null'. This accommodates for situations when Product collection block is first added to the page.

2. Various improvements in 'index.tsx' file which include more robust null checking for the 'query' object, simplifying the way 'woocommerceAttributes' is obtained, and passing 'setQueryAttributeBind' and 'inherit' to the 'InheritQueryControl' component.

3. In 'inherit-query-control.tsx', 'InheritQueryControl' component has been refactored to use '__experimentalToolsPanelItem' from '@wordpress/components' instead of 'ToggleControl'. This adds more flexibility to the control.

4. Changes in 'ProductCollectionAttributes' and 'ProductCollectionQuery' types in 'types.ts'. The

imanish003 committed:
Resolve conflicts

imanish003 committed:
Improve product collection query inheritance and fix URL typo

This commit addresses two primary areas:
1. Fixed a typo in the URL used as a reference in the use-previous.ts file. The URL was incorrectly case-sensitive, which has been corrected.

2. The product-collection block in the JavaScript files has been refactored for better handling of query inheritance:

   - Changed the 'inherit' value from false to null in the DEFAULT_QUERY constant to handle initial state more accurately.

   - In product-collection/inspector-controls, implemented conditional rendering for Filters and Query Controls based on 'displayQueryControls'. Also, improved the 'InheritQueryControl' by using the 'usePrevious' hook to maintain the state before enabling the inheritance.

   - In inherit-query-control, enhanced the control to toggle the query inheritance. It now considers the 'inherit' state from the query object and keeps track of the query object state before enabling inheritance. If the inheritance is toggled off, it reverts the query to the previous state before inheritance was enabled.

imanish003 committed:
Resolve conflicts

imanish003 committed:
Minor improvemnets

imanish003 committed:
Merge branch 'trunk' into add/product-collection-inherit

imanish003 commented:
Hey @Aljullu, I've assigned this PR to you for review since @albarin is currently AFK. I hope that's alright. 🙂

PS: I have also updated the description to reflect the new changes made in this PR.

imanish003 commented:
@kmanijak 

> When adding the Product Collection, attributes are correctly set in the Editor (client-side rendered), but when you save with no changes and go to frontend then query is not consumed. This is due to fact that until any change is made in the inspector control and until setAttributes function is called the attributes are not saved to the block and hence not read by server-side.

FYI I found [this issue](https://github.com/WordPress/gutenberg/issues/7342) on GB repo which explains why this was happening. I fixed the issue using following code:
https://github.com/woocommerce/woocommerce-blocks/blob/49f4fefbb45da769cc656672e3123cfeb29e8dc2/assets/js/blocks/product-collection/edit.tsx#L87-L93

imanish003 committed:
Resolve conflicts

imanish003 committed:
Add wc-block-editor- prefix with classNames

imanish003 committed:
Handle duplicate taxonomy names in Taxonomy controls

the taxonomy controls have been enhanced in the following ways:
1. Modified the BASE_QUERY object to include 'slug' in the '_fields' property. This will ensure that the 'slug' of the taxonomy term is fetched along with its 'id' and 'name'.

2. Added a 'slug' property to the Term type to store the 'slug' of each term.

3. Updated the useEffect hook inside the TaxonomyItem function to generate suggestions based on search results. The suggestions now include the 'slug' of a term if the term's name is not unique. This change will help users distinguish between terms with the same name but different slugs.

imanish003 committed:
Merge branch 'trunk' into 9364-product-collection-filters-taxonomies

imanish003 committed:
Remove isset() in if condition as it's unnecessary

imanish003 committed:
Refactor TaxonomyItem component for better readability

Following changes were made:

1. The useSelect hooks which were being used to fetch existing terms and search results have been moved into their own custom hooks named 'useExistingTerms' and 'useSearchResults' respectively. This simplifies the TaxonomyItem function's body and makes the hooks' purposes clearer.

2. The comments and props destructuring for the TaxonomyItem function have been moved up to make it easier to understand the function's purpose and the props it receives.

3. This refactor does not introduce any changes in functionality. It only changes how the code is organized and presented, which will make future development easier.

imanish003 committed:
Handling for duplicate term names & other improvements

This commit enhances the `TaxonomyControls` component within `product-collection` block by adding memoization and improving term uniqueness handling.

Changes:

1. Imported `useMemo` from `@wordpress/element` for memoizing certain results.

2. `getTermIdByTermValue` function has been modified to use a `termIdToNameMap` (term ids as keys and term names as values). This provides a more efficient and direct mapping for term search.

3. Introduced `useTermIdToNameMap` function, which returns a `Map` where term ids are keys and term names are values. It handles duplicate term names by appending the term slug to the name, ensuring unique term names.

4. Updated the `useExistingTerms` and `useSearchResults` to include `taxonomy` in their dependency arrays for `useSelect` hook. This will force re-computation when `taxonomy` changes.

5. Changed `TaxonomyItem` from a function declaration to a const arrow function, consistent with the rest of the codebase.

6. Updated `onTermsChange` function in `TaxonomyItem` to accommodate the changes in `getTermIdByTermValue` and the introduction of `termIdToNameMap`.

7. Replaced `Set` with a standard array for storing new term IDs in `onTermsChange`. The `Set` was unnecessary as term IDs are unique by default.

8. Updated `TaxonomyItem`'s effects and rendering to work with `termIdToNameMap`, ensuring the displayed term names are unique.

This update will result in more efficient term search and handling, and it will solve issues related to duplicate term names.

imanish003 committed:
Restructure taxonomy controls in product collection block

This commit restructures the taxonomy controls in the product collection block for improved clarity and maintainability.
- The file `taxonomy-controls.tsx` has been deleted, and its functionality has been divided into two new files: `index.tsx` and `taxonomy-item.tsx`.

- The `index.tsx` file contains the main TaxonomyControls component, which is responsible for displaying taxonomy-related options in the block's inspector controls. It includes a custom hook `useTaxonomies` that fetches and returns taxonomies associated with product post type.

- The `taxonomy-item.tsx` file, on the other hand, contains a TaxonomyItem component that handles the rendering of individual taxonomy items. It also contains some utility functions for mapping term names and ids and fetching terms based on the search query.

This refactor aims to improve code readability and separation of concerns, thus making future changes and maintenance easier.

imanish003 committed:
Merge branch 'trunk' into 9364-product-collection-filters-taxonomies

imanish003 committed:
Merge branch '9364-product-collection-filters-taxonomies' of https://github.com/woocommerce/woocommerce-blocks into add/product-collection-inherit

imanish003 committed:
Fix case insensitive search support for FormTokenField

This change enhances the search functionality of the FormTokenField by introducing support for case insensitive search. This has been achieved by adding a lower-case version of the term name to the 'termNameToIdMap'.

This is an important enhancement as it will make the search process more user-friendly and resilient to different casing inputs. Users will now be able to find the desired taxonomy term regardless of their input's case.

imanish003 committed:
Refactor getTermIdByTermValue function and update newSuggestions mapping

This commit does a couple of important things:
1. Reorders the definition of constants in `TaxonomyItemProps` for clarity.
2. Refactors the `getTermIdByTermValue` function. Instead of checking for an exact term name match in a convoluted manner, it now directly tries to fetch the `id` from the `searchTerm` if it is an object. If the `searchTerm` is not an object, the function tries to match it against the `termNameToIdMap` in both normal and lowercase forms. This simplification makes the function more readable and concise.
3. Updates the `newSuggestions` mapping in the `TaxonomyItem` component. It now has a fallback to `searchResult.name` if a term's name is not found in `termIdToNameMap`. This change ensures that even if the term's name is not in the map for some reason, we can still display a suggestion using the original name of the term.

imanish003 committed:
Merge branch 'trunk' into 9364-product-collection-filters-taxonomies

imanish003 committed:
Optimize term fetching and initial search state in TaxonomyItem

This commit introduces a couple of improvements to the TaxonomyItem component.

1. The initial state of the 'search' state variable has been updated to 'undefined'. This change helps prevent unnecessary initial fetching of terms when the search input is empty.

2. Term fetching logic has been optimized to only enable term fetching when necessary:
   a) Fetching based on the search query is only enabled when 'search' is not 'undefined'.
   b) Fetching existing terms is only enabled when there are term IDs.

3. The block of code responsible for fetching existing terms and setting the current value has been moved upwards. This reordering of code does not change the functionality, but it groups together similar pieces of code, enhancing readability and maintainability.

These optimizations make the component more efficient by reducing unnecessary requests and computations, and they improve the code organization.

imanish003 committed:
Merge branch '9364-product-collection-filters-taxonomies' into add/product-collection-inherit

imanish003 committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into add/product-collection-inherit

Aljullu commented on assets/js/blocks/product-collection/inspector-controls/columns-control.tsx:
```diff
@@ -22,6 +22,8 @@ import { getDefaultSettings } from '../constants';
 const ColumnsControl = (
 	props: BlockEditProps< ProductCollectionAttributes >
 ) => {
+	if ( ! props.attributes.displayLayout ) return null;
```
I'm a bit confused by this change, does it have any relation to the rest of the PR or is it a generic code cleanup? (Sorry if it might seem obvious, but wanted to make sure I'm not missing anything)

As a side note, I noticed the `ColumnsControl` doesn't have any title or label. Is that on purpose? I feel it's a bit confusing to have a range input field without any info about what it does:

![imatge](https://github.com/woocommerce/woocommerce-blocks/assets/3616980/2084b4ec-2ede-4aee-94cb-fb3560a4c25f)

(No need to fix it in this PR, but wanted to mention it anyway :slightly_smiling_face: )

imanish003 replied:
> I'm a bit confused by this change, does it have any relation to the rest of the PR or is it a generic code cleanup? (Sorry if it might seem obvious, but wanted to make sure I'm not missing anything)

Actually, there was an issue where the `displayLayout` sometimes became null. Therefore, I added an early return statement. However, I have now removed it in [34762c5](https://github.com/woocommerce/woocommerce-blocks/pull/9485/commits/34762c5190f87382370bcec12177ae0416f44299) as it's no longer necessary due to other changes made in the commit.

I truly appreciate your thorough PR reviews. 🙇‍♂️

imanish003 replied:
> As a side note, I noticed the ColumnsControl doesn't have any title or label. Is that on purpose? I feel it's a bit confusing to have a range input field without any info about what it does:

I have also added a missing label in [34762c5](https://github.com/woocommerce/woocommerce-blocks/pull/9485/commits/34762c5190f87382370bcec12177ae0416f44299). Thanks for pointing this out 🙌🏻

Aljullu commented on assets/js/blocks/product-collection/inspector-controls/index.tsx:
```diff
@@ -46,33 +51,43 @@ const ProductCollectionInspectorControls = (
+						setQueryAttribute={ setQueryAttributeBind }
+						query={
+							props.attributes.query as ProductCollectionQuery
+						}
```
Nit: this could be changed to `query`, now that we define it in line 33, no? Also, on the first render after inserting the block, `query` might be undefined, so I would remove or update the type (`as ProductCollectionQuery`). What do you think?

imanish003 replied:
Updated, Thanks for pointing this out 🙌🏻

Aljullu commented on assets/js/blocks/product-collection/inspector-controls/inherit-query-control.tsx:
```diff
@@ -0,0 +1,112 @@
+		? ARCHIVE_PRODUCT_TEMPLATES.includes( currentTemplateId )
+		: false;
+	if ( inherit === null ) {
+		setQueryAttribute( {
+			inherit: initialValue,
+		} );
```
Sometimes, when adding the Product Collection block into the Site Editor, I get this JS warning in the console:

> Warning: Cannot update a component (`DocumentActions`) while rendering a different component (`InheritQueryControl`). To locate the bad setState() call inside `InheritQueryControl`, follow the stack trace as described in...

Can you reproduce as well?

I think the error is caused by this `setQueryAttribute()`. If I'm understanding the code correctly, it's trying to update the state during the render function, which causes an unnecessary re-render and might cause other side effects. I think we should set the value of `inherit` to `initialValue` right when we read it, so in `assets/js/blocks/product-collection/inspector-controls/index.tsx`. Would that work? :thinking:

imanish003 replied:
I can also see this warning. I have relocated the code responsible for calculating the inherit value to:
[edit.tsx](https://github.com/woocommerce/woocommerce-blocks/blob/34762c5190f87382370bcec12177ae0416f44299/assets/js/blocks/product-collection/edit.tsx#L96-L119).

Now, I am setting the inherit value while defining the default attributes. Additionally, I have made some other minor improvements in [34762c5](https://github.com/woocommerce/woocommerce-blocks/pull/9485/commits/34762c5190f87382370bcec12177ae0416f44299).

What are your thoughts on the recent changes? Do they look good to you?

Aljullu replied:
Yup, looks good. I initially thought the `useEffect()` would not be necessary, but I guess because of https://github.com/WordPress/gutenberg/issues/7342, there is no way we can avoid it. :disappointed:

Aljullu reviewed (commented):
This is testing as expected and the code looks overall good, but I left some questions/comments below: :point_down:

imanish003 committed:
Address PR comments and other improvements

This commit makes several changes:
1. The useEffect that sets the default attributes was moved and modified. This now includes a `query` attribute that utilizes the imported function `getDefaultValueOfInheritQueryFromTemplate`.
2. An early return was added in `edit.tsx` to prevent rendering until default attributes are set.
3. In `columns-control.tsx`, the early return was removed and a label was added to the `RangeControl` component.
4. In `inherit-query-control.tsx`, logic related to `inherit` value initial setting was refactored using a `useMemo` hook with `getDefaultValueOfInheritQueryFromTemplate` function. This logic was moved to a separate utility function in `utils`.
5. The `query` attribute is no longer optional in `types.ts`.
6. A new utility function `getDefaultValueOfInheritQueryFromTemplate` was created in `utils.tsx` to encapsulate the logic of deciding the default value of `inherit` query attribute based on the current template.

These changes aim to improve code clarity and maintainability.

imanish003 commented:
@Aljullu I have made the changes. Can you please take a look? 🙂

Also, I am immensely grateful for yet another remarkable PR review you have provided. Your consistent delivery of exceptional feedback is truly appreciated. Thank you for consistently going above and beyond in providing such outstanding feedback. 🙌🏻 🙇‍♂️

Aljullu commented on assets/js/blocks/product-collection/edit.tsx:
```diff
@@ -10,11 +10,12 @@ import { useEffect } from '@wordpress/element';
  */
 import { ImageSizing } from '../../atomic/blocks/product-elements/image/types';
-import { ProductCollectionAttributes } from './types';
+import { ProductCollectionAttributes, ProductCollectionQuery } from './types';
```
Nit: should we intentionally use `import type` here?
```suggestion
import type { ProductCollectionAttributes, ProductCollectionQuery } from './types';
```

imanish003 replied:
Although I don't have a strong opinion about this so I will go ahead and add the `type` with import. It might help in clarifying the intent. 🤝

Aljullu reviewed (approved):
Thanks for all the improvements, @imanish003. It looks good to me! :ship:

imanish003 committed:
Add  with types import statement

imanish003 committed:
Merge branch 'trunk' into add/product-collection-inherit
